core/threaded/watchdog.c

```diff
@@ -2,82 +2,132 @@
+    int ret;
+    for (int i = 0; i < env->watchdogs_size; i++) {
+        watchdog_t *watchdog = env->watchdogs[i];
+        if (watchdog->thread_active) {
+            ret = lf_mutex_lock(watchdog->base->reactor_mutex);
+            LF_ASSERTN(ret, "lf_mutex_lock failed");
+            lf_watchdog_stop(watchdog);
+            ret = lf_mutex_unlock(watchdog->base->reactor_mutex);
+            LF_ASSERTN(ret, "lf_mutex_unlock failed");
+            ret = lf_thread_join(watchdog->thread_id, &thread_return);
+            LF_ASSERTN(ret, "lf_thread_join failed");
         }
     }
 }
 
 /**
  * @brief Thread function for watchdog.
- * This function sleeps until physical time exceeds the expiration time of
- * the watchdog and then invokes the watchdog expiration handler function.
+ * Each watchdog has a thread which sleeps until one out of two scenarios:
+ * 1) The watchdog timeout expires and there has not been a renewal of the watchdog budget.
+ * 2) The watchdog is signaled to wake up and terminate.
  * In normal usage, the expiration time is incremented while the thread is
- * sleeping, so the watchdog never expires and the handler function is never
- * invoked.
- * This function acquires the reaction mutex and releases it while sleeping.
+ * sleeping, so when the thread wakes up, it can go back to sleep again.
+ * If the watchdog does expire. It will execute the watchdog handler and the 
+ * thread will terminate. To stop the watchdog, another thread will signal the
+ * condition variable, in that case the watchdog thread will terminate directly.
+ * The expiration field of the watchdog is used to protect against race conditions.
+ * It is set to NEVER when the watchdog is terminated.
  * 
  * @param arg A pointer to the watchdog struct
  * @return NULL
  */
 void* _lf_run_watchdog(void* arg) {
+    int ret;
     watchdog_t* watchdog = (watchdog_t*)arg;
+    LF_PRINT_DEBUG("Starting Watchdog %p", watchdog);
 
     self_base_t* base = watchdog->base;
-    assert(base->reactor_mutex != NULL);
-    lf_mutex_lock((lf_mutex_t*)(base->reactor_mutex));
+    LF_ASSERT(base->reactor_mutex, "reactor-mutex not alloc'ed but has watchdogs.");
+    ret = lf_mutex_lock((lf_mutex_t*)(base->reactor_mutex));
+    LF_ASSERTN(ret, "lf_mutex_lock failed");
     instant_t physical_time = lf_time_physical();
-    while (physical_time < watchdog->expiration) {
+    while (watchdog->expiration != NEVER && physical_time < watchdog->expiration) {
         interval_t T = watchdog->expiration - physical_time;
-        lf_mutex_unlock((lf_mutex_t*)base->reactor_mutex);
-        lf_sleep(T);
-        lf_mutex_lock((lf_mutex_t*)(base->reactor_mutex));
+        LF_PRINT_DEBUG("Watchdog %p going to sleep", watchdog);
+        // Wait for expiration, or a signal to terminate
+        if(lf_cond_timedwait(&watchdog->cond, watchdog->expiration) != LF_TIMEOUT) {
+            LF_PRINT_DEBUG("Watchdog %p was cancelled. Terminating", watchdog);
+            goto terminate;
+        }
```

Comment:
This looks dangerous to me. `lf_cond_timedwait` uses `pthreads_cond_timedwait` which is, according to the documentation, subject to spurious wake ups.  This could cause the watchdog to randomly appear to have been cancelled when there was no intent to cancel.  I think the cancelling has to depend on more than the return value of `lf_cond_timedwait`.  Would it be sufficient to check that `watchdog->expiration == NEVER`?

Reply:
Thanks addressed now